File: 200. Number of Islands/main.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function (grid) {
  if (!grid || grid.length === 0) return 0;

  const m = grid.length;
  const n = grid[0].length;

  const visited = new Set();
  let numOfIslands = 0;

  for (let r = 0; r < m; ++r) {
    for (let c = 0; c < n; ++c) {
      const key = `${r},${c}`;
      if (grid[r][c] === "0" || visited.has(key)) continue;

      ++numOfIslands;
      dfs(r, c);
    }
  }

  return numOfIslands;

  function dfs(r, c) {
    if (r < 0 || c < 0 || r >= m || c >= n) return;
    if (grid[r][c] === "0") return;

    const key = `${r},${c}`;
    if (visited.has(key)) return;
    visited.add(key);

    dfs(r + 1, c);
    dfs(r - 1, c);
    dfs(r, c + 1);
    dfs(r, c - 1);
  }
};

File: 200. Number of Islands/main.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
    
        def bfs(i,j):

            if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]):
                return

            if grid[i][j] == "0" or ((i,j) in visited):
                return
            
            visited.add((i,j))

            bfs(i - 1, j)
            bfs(i + 1, j)
            bfs(i, j - 1)
            bfs(i, j + 1)

        numOfIslands = 0
        visited = set()

        for i in range(len(grid)):
            for j in range(len(grid[i])):

                if grid[i][j] == "0":
                    continue
                if (i,j) in visited:
                    continue

                bfs(i,j)
                numOfIslands += 1

        return numOfIslands
                
                